### PR TITLE
IGNITE-22564 Decouple metastore and cmg RAFT log from data partitions…

### PR DESCRIPTION
**Pull request by Cyrill**
https://issues.apache.org/jira/browse/IGNITE-22564

### CODE REVIEW

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java [hunk 1]
```diff
@@ -202,9 +196,14 @@ public <T extends RaftGroupService> CompletableFuture<T> startRaftGroupNode(
             PeersAndLearners configuration,
             RaftGroupListener lsnr,
             RaftGroupEventsListener eventsLsnr,
-            RaftServiceFactory<T> factory
+            RaftServiceFactory<T> factory,
+            RaftOptionsConfigurator storageConfigurator
```
Comment (vldpyatkov) [inline modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java:200]: Why did you choose a closure instead of pojo with clear fields?

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java [hunk 2]
```diff
@@ -314,25 +325,29 @@ public <T extends RaftGroupService> CompletableFuture<T> startRaftGroupNodeAndWa
...
             throw new NodeStoppingException();
         }
 
         try {
-            CompletableFuture<T> startRaftServiceFuture = startRaftGroupNodeInternal(
+            RaftGroupOptions raftGroupOptions = RaftGroupOptions.defaults();
+
+            storageConfigurator.configure(raftGroupOptions);
+
+            raftGroupOptions.ownFsmCallerExecutorDisruptorConfig(disruptorConfiguration);
```
Comment (vldpyatkov) [inline modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java:340]: Why don't we reduce this method from Loza and add this setter to the raft option configuration closure (msRaftConfigurator)?

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 3]
```diff
@@ -855,6 +862,8 @@ private RaftGroupOptions groupOptionsForPartition(boolean isVolatileStorage, Sna
 
         raftGroupOptions.commandsMarshaller(raftCommandsMarshaller);
 
+        partitionRaftConfigurator.configure(raftGroupOptions);
```
Comment (vldpyatkov) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:865]: Only Loza do this in code above. Why do you think it ought to be done explicitly before Loza is called?

**Review by vldpyatkov — APPROVED** [review]

File: modules/metastorage/src/integrationTest/java/org/apache/ignite/internal/metastorage/impl/ItIdempotentCommandCacheTest.java [hunk 4]
```diff
@@ -179,11 +195,22 @@ private static class Node implements AutoCloseable {
...
+            msLogStorageFactory =
+                    SharedLogStorageFactoryUtils.create(clusterService.nodeName(), metastorageWorkDir.raftLogPath());
+
+            RaftOptionsConfigurator msRaftConfigurator = options -> {
+                RaftGroupOptions raftOptions = (RaftGroupOptions) options;
+
+                // TODO: use interface, see https://issues.apache.org/jira/browse/IGNITE-18273
+                raftOptions.setLogStorageFactory(msLogStorageFactory);
+                raftOptions.serverDataPath(metastorageWorkDir.metaPath());
+            };
```
Comment (rpuch) [inline modules/metastorage/src/integrationTest/java/org/apache/ignite/internal/metastorage/impl/ItIdempotentCommandCacheTest.java:209]: This spell seems to be repeated a lot. How about extracting a method with 2 parameters and using it instead?

File: modules/metastorage/src/integrationTest/java/org/apache/ignite/internal/metastorage/impl/ItMetaStorageMultipleNodesAbstractTest.java [hunk 5]
```diff
@@ -130,6 +135,12 @@ private class Node {
 
         private final Loza raftManager;
 
+        private final LogStorageFactory defaultLogStorageFactory;
+
+        private final LogStorageFactory msLogStorageFactory;
+
+        private final LogStorageFactory cmgLogStorageFactory;
```
Comment (rpuch) [inline modules/metastorage/src/integrationTest/java/org/apache/ignite/internal/metastorage/impl/ItMetaStorageMultipleNodesAbstractTest.java:142]: Why are these different? Should they both be the same factory?

File: modules/placement-driver/build.gradle [hunk 6]
```diff
@@ -57,6 +57,7 @@ dependencies {
     integrationTestImplementation project(':ignite-transactions')
     integrationTestImplementation project(':ignite-catalog')
     integrationTestImplementation project(':ignite-metrics')
+    integrationTestImplementation project(':ignite-runner')
```
Comment (rpuch) [inline modules/placement-driver/build.gradle:60]: Why is this needed here?

File: modules/raft/src/integrationTest/java/org/apache/ignite/raft/server/ItJraftServerLogPathTest.java [hunk 7]
```diff
@@ -78,18 +81,6 @@ void testDefaultFactory() {
         assertTrue(Files.exists(partitionsPath));
     }
 
-    @Test
-    @WithSystemProperty(key = SharedLogStorageFactoryUtils.LOGIT_STORAGE_ENABLED_PROPERTY, value = "true")
```
Comment (rpuch) [inline modules/raft/src/integrationTest/java/org/apache/ignite/raft/server/ItJraftServerLogPathTest.java:82]: Why is this test removed? Is it moved elsewhere?

File: modules/raft-api/src/main/java/org/apache/ignite/internal/raft/RaftOptionsConfigurator.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+ *        RaftGroupOptions raftOptions = (RaftGroupOptions) options;
+ *
+ *        raftOptions.setLogStorageFactory(logStorageFactory);
+ *        raftOptions.serverDataPath(dataPath);
+ *    };
+ * </pre>
+ * TODO: https://issues.apache.org/jira/browse/IGNITE-18273
+ */
+@FunctionalInterface
+public interface RaftOptionsConfigurator {
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/raft-api/src/main/java/org/apache/ignite/internal/raft/RaftOptionsConfigurator.java:35]: >The name comes from the action it performs - to configure raft parameters
It configures group options, not overall raft options.

There is a class called exactly `RaftOptions`, and the proposed naming creates confusion. Everyone who knows about the implementation (that is, most AI3 developers) would think about this class first.

File: modules/metastorage/src/integrationTest/java/org/apache/ignite/internal/metastorage/impl/ItMetaStorageMultipleNodesAbstractTest.java (hunk 5 above)
Comment (rpuch) [reply modules/metastorage/src/integrationTest/java/org/apache/ignite/internal/metastorage/impl/ItMetaStorageMultipleNodesAbstractTest.java:142]: But why do we need this?

File: modules/raft/src/main/java/org/apache/ignite/internal/configuration/RaftOptionsConfigurationHelper.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    /**
+     * Creates an instance of {@link RaftOptionsConfigurator} that can be used to configure Raft options.
+     *
+     * @param logStorageFactory Log storage factory.
+     * @param serverDataPath Server data path.
+     * @return Raft options configurator.
+     */
+    static RaftOptionsConfigurator configureProperties(LogStorageFactory logStorageFactory, Path serverDataPath) {
+        return options -> {
+            RaftGroupOptions raftOptions = (RaftGroupOptions) options;
```
Comment (rpuch) [inline modules/raft/src/main/java/org/apache/ignite/internal/configuration/RaftOptionsConfigurationHelper.java:39]:
```suggestion
            RaftGroupOptions groupOptions = (RaftGroupOptions) options;
```

File: modules/raft-api/src/main/java/org/apache/ignite/internal/raft/RaftOptionsConfigurator.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/raft-api/src/main/java/org/apache/ignite/internal/raft/RaftOptionsConfigurator.java:35]: >But I can replace it with Configurer, if you insist.

It's up to you, it seems to be a stylistic matter

File: modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java [hunk 10]
```diff
@@ -548,16 +548,8 @@ public class IgniteImpl implements Ignite {
...
-            RaftGroupOptions raftOptions = (RaftGroupOptions) options;
-            // TODO: use interface, see https://issues.apache.org/jira/browse/IGNITE-18273
-            if (raftOptions.getLogStorageFactory() == null) {
-                raftOptions.setLogStorageFactory(cmgLogStorageFactory);
-            }
-            if (raftOptions.serverDataPath() == null) {
-                raftOptions.serverDataPath(cmgWorkDir.metaPath());
-            }
-        };
+        RaftGroupOptionsConfigurer cmgRaftConfigurator =
```
Comment (rpuch) [inline modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java:551]: Type says that it's configurer, but variable is a configurator. Should they agree?

**Review by rpuch — APPROVED** [review]

**Review by ibessonov — APPROVED** [review]
